### DISCUSSION TITLE
A few questions to UNION

### DISCUSSION THREAD

**Discussion by ghost** [jOOQ/jOOQ#15924] [Q&A] [answered]
Hi,

I have a different question in regards to UNIONs this time. Is it true that .union() only accepts subclasses of Record? So... code generation again?

Also: I would like to extract/change an existing union statement. It seems like the union-field is also private in SelectQueryImpl. Is there some to access it? Perhaps through the QOM-API?

I also didn't quite understand why the union-Field has Lists of Lists of select statements. Are these all supposed to be unions too?

Thanks!

**Comment by lukaseder** [comment, marked as answer]
> I have a different question in regards to UNIONs this time. Is it true that .union() only accepts subclasses of Record? So... code generation again?

I'm not sure what you mean by this, but I think the method signature is clear? The two union subqueries must have matching row types. This has nothing to do with code generation.

> Also: I would like to extract/change an existing union statement. It seems like the union-field is also private in SelectQueryImpl. Is there some to access it? Perhaps through the QOM-API?

Unfortunately, `UNION` (and other set operations) aren't yet available from the `QOM` API as of jOOQ 3.19, see:

- https://github.com/jOOQ/jOOQ/issues/15906

> I also didn't quite understand why the union-Field has Lists of Lists of select statements. Are these all supposed to be unions too?

I'm assuming you're talking about the implementation, because there's no such type in the API that I'm aware of. The `SelectQueryImpl` can contain several set operation, and it pretends that a set operator isn't just a binary operator (e.g. `(S1 UNION S2 UNION S3) EXCEPT S4`, where `S1, S2, S3` are a list of selects at position `0` and 
`S4` is a list of selects at position 1)

**Comment by lukaseder** [comment]
> I'm not sure what you mean by this, but I think the method signature is clear? The two union subqueries must have matching row types. This has nothing to do with code generation.

I realised that you may have concluded from the signature that the expected `Record` subtype is e.g. a generated `BookRecord` or whatever. But the main strength of passing around `R extends Record` all over the place in jOOQ is not that, it's being type safe for tuple types, such as:

```java
// Type checks:
select(T.INTEGER_COLUMN, T.STRING_COLUMN).from(T)
.union(select(U.INTEGER_COLUMN, U.STRING_COLUMN).from(U));

// Doesn't type check (Record2<Integer, String> vs Record2<String, Integer>)
select(T.INTEGER_COLUMN, T.STRING_COLUMN).from(T)
.union(select(U.STRING_COLUMN, U.INTEGER_COLUMN).from(U));

// Doesn't type check (Record2<Integer, String> vs Record1<Integer>)
select(T.INTEGER_COLUMN, T.STRING_COLUMN).from(T)
.union(select(U.INTEGER_COLUMN).from(U));
```
